app/src/org/commcare/activities/connect/ConnectIDNetworkHelper.java

```diff
@@ -0,0 +1,267 @@
+import org.json.JSONObject;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+
+import okhttp3.MediaType;
+import okhttp3.RequestBody;
+
+public class ConnectIDNetworkHelper {
+    public interface INetworkResultHandler {
+        void processSuccess(int responseCode, InputStream responseData);
+        void processFailure(int responseCode, IOException e);
+    }
+
+    public static void post(Context context, String url, AuthInfo authInfo, HashMap<String, String> params, INetworkResultHandler handler) {
+        Gson gson = new Gson();
+        String json = gson.toJson(params);
+
+        RequestBody requestBody = RequestBody.create(MediaType.parse("application/json"), json);
+        ModernHttpTask postTask =
+                new ModernHttpTask(context, url,
+                        ImmutableMultimap.of(),
+                        new HashMap<>(),
+                        requestBody,
+                        HTTPMethod.POST,
+                        authInfo);
+        postTask.connect(new ConnectorWithHttpResponseProcessor<>() {
+            @Override
+            public void processSuccess(int responseCode, InputStream responseData) {
+                handler.processSuccess(responseCode, responseData);
+            }
+
+            @Override
+            public void processClientError(int responseCode) {
+                //400 error
+                handler.processFailure(responseCode, null);
+            }
+
+            @Override
+            public void processServerError(int responseCode) {
+                //500 error for internal server error
+                handler.processFailure(responseCode, null);
+            }
+
+            @Override
+            public void processOther(int responseCode) {
+                handler.processFailure(responseCode, null);
+            }
+
+            @Override
+            public void handleIOException(IOException exception) {
+                //UnknownHostException if host not found
+                handler.processFailure(-1, exception);
+            }
+
+            @Override
+            public <A, B, C> void connectTask(CommCareTask<A, B, C, HttpResponseProcessor> task) {}
+
+            @Override
+            public void startBlockingForTask(int id) {}
+
+            @Override
+            public void stopBlockingForTask(int id) {}
+
+            @Override
+            public void taskCancelled() {}
+
+            @Override
+            public HttpResponseProcessor getReceiver() { return this; }
+
+            @Override
+            public void startTaskTransition() {}
+
+            @Override
+            public void stopTaskTransition(int taskId) {}
+
+            @Override
+            public void hideTaskCancelButton() {}
+        });
+        postTask.executeParallel();
+    }
+
+    public static void post(Context context, String url, AuthInfo authInfo, Multimap<String, String> params, INetworkResultHandler handler) {
```

Comment:
we should be able to use this as the only `post` method here and always pass params as multimap.

Reply:
These two post() functions are actually sending data differently. The first version encodes the params input as JSON in the body, while the second uses form encoding. I use the first call for everything going to the ConnectID server EXCEPT for OAuth, which requires the second call. Perhaps the functions should have different names to be more clear, i.e. postWithJson and postWithFormEncoding?

Reply:
Done.